job_creator.php

```diff
@@ -75,6 +79,51 @@ public function createJob(int $phpIndex, array $opts): array
         return array_merge($default, $opts);
     }
 
+    private function isAllowedPhpVersion(string $phpVersion)
+    {
+        $phpVersion = (float) $phpVersion;
+        // no composer.json file or php version no defined in composer.json, just allow the php version
+        if ($this->composerPhpConstraint === null) {
+            return true;
+        }
+        if (!file_exists($this->composerJsonPath)) {
+            $this->composerPhpConstraint = null;
+            return true;
+        }
+        $json = json_decode(file_get_contents($this->composerJsonPath));
+        if (!isset($json->require->php)) {
+            $this->composerPhpConstraint = null;
+            return true;
+        }
+        if ($this->composerPhpConstraint === '') {
+            $this->composerPhpConstraint = $json->require->php;
+        }
+        $constraints = explode('||', $this->composerPhpConstraint);
+        $constraints = array_map(function($php) {
+            return str_replace(' ', '', $php);
+        }, $constraints);
+        foreach ($constraints as $constraint) {
+            $composerVersion = preg_replace('#[^0-9\.]#', '', $constraint);
+            // convert semver php versions to minors, github actions will also just use the latest patch so
+            // we can assume that ignoring the patch versio is safe
+            $composerVersion = preg_replace('#^([0-9]+)\.([0-9]+)\.[0-9]$#', '$1.$2', $composerVersion);
+            $composerVersion = (float) $composerVersion;
+            $op = preg_replace('#[0-9\.]#', '', $constraint);
+            if (
+                ($op == '^' && floor($phpVersion) == floor($composerVersion) && $phpVersion >= $composerVersion) ||
+                ($op == '~' && $phpVersion == $composerVersion) ||
```

Comment:
We do actually need to know if `$composerVersion` had a patch version here. 
`~1.2.3` means `>=1.2.3, < 1.3.0` - doesn't allow higher minor versions.
`~1.2` is functionally the same as `^1.2` as it allows higher minor versions.
From [the docs](https://getcomposer.org/doc/articles/versions.md#tilde-version-range-)
> `~1.2` is equivalent to `>=1.2 <2.0.0`, while `~1.2.3` is equivalent to `>=1.2.3 <1.3.0`